File: kernel/livepatch/dummy_target.rs
// SPDX-License-Identifier: GPL-2.0-only
// Copyright (C) 2026 rezky_nightky

//! Dummy kernel target for safe livepatch testing.
//!
//! Provides a simple, isolated function that can be safely patched.
//! Zero risk to the core kernel — this function only prints a trace
//! and returns a known value. The patch replaces it with a new
//! function that returns a different value.

use alloc::vec::Vec;

/// Known return value of the unpatched function.
pub const DUMMY_ORIGINAL_VALUE: u64 = 42;

/// Known return value of the patched function.
pub const DUMMY_PATCHED_VALUE: u64 = 99;

/// The dummy target function — safe to patch.
///
/// Returns `DUMMY_ORIGINAL_VALUE` (42) when unpatched.
/// After livepatch, returns `DUMMY_PATCHED_VALUE` (99).
///
/// In a real kernel module, this would be:
/// ```c
/// // Marked __visible so it can be patched via ftrace
/// u64 __visible zenvecha_dummy_func(void) {
///     return 42;
/// }
/// ```
pub fn zenvecha_dummy_func() -> u64 {
    // In real kernel code, this function is compiled with
    // __attribute__((__section__(".text.zenvecha"))) and
    // NOTRACE / NOKPROBE annotations to prevent other
    // tracing frameworks from interfering.
    DUMMY_ORIGINAL_VALUE
}

/// The replacement function — patched over the dummy.
///
/// This is the "new function" payload. After livepatch,
/// calls to `zenvecha_dummy_func` are redirected here.
///
/// Returns `DUMMY_PATCHED_VALUE` (99) instead of 42.
pub fn zenvecha_dummy_patched() -> u64 {
    DUMMY_PATCHED_VALUE
}

/// Verify the patch by calling the target function and
/// checking the return value.
pub fn verify_patch() -> bool {
    zenvecha_dummy_func() == DUMMY_PATCHED_VALUE
}

/// Dummy target metadata exposed via proc.
pub struct DummyTargetInfo {
    pub name: &'static str,
    pub address: usize,
    pub original_value: u64,
    pub patched_value: u64,
    pub is_patched: bool,
}

impl DummyTargetInfo {
    pub fn new() -> Self {
        DummyTargetInfo {
            name: "zenvecha_dummy_func",
            address: zenvecha_dummy_func as *const () as usize,
            original_value: DUMMY_ORIGINAL_VALUE,
            patched_value: DUMMY_PATCHED_VALUE,
            is_patched: verify_patch(),
        }
    }
}

File: kernel/dummy_target.c
// SPDX-License-Identifier: GPL-2.0-only
// Copyright (C) 2026 rezky_nightky

//! Dummy kernel target for safe livepatch testing.
//!
//! Provides a simple, isolated function that can be safely patched.
//! Zero risk to the core kernel — this function only returns a known
//! value. The patch flips an atomic flag (via stop_machine) that
//! switches the return value.
//!
//! ## Skeleton vs Production
//!
//! Skeleton (current):
//!   - zenvecha_dummy_func() checks atomic flag, returns 42 or 99
//!   - "Patching" = stop_machine + atomic_set(flag, 1)
//!   - "Reverting" = stop_machine + atomic_set(flag, 0)
//!   - Proves the atomicity model without modifying kernel text
//!
//! Production (future):
//!   - zenvecha_dummy_func marked __visible + NOTRACE
//!   - Patch installs ftrace handler at the function entry
//!   - Calls are redirected to zenvecha_dummy_patched()
//!   - stop_machine ensures no CPU is mid-execution during the swap

#include <linux/kernel.h>
#include <linux/atomic.h>
#include <linux/string.h>
#include "zenvecha.h"

static atomic_t zenvecha_dummy_patched = ATOMIC_INIT(0);

const char *zenvecha_dummy_target_name(void)
{
	return "zenvecha_dummy_func";
}

/* The dummy target function.
 *
 * Marked noinline to prevent the compiler from inlining it into
 * callers (which would break the patching contract). The notrace
 * annotation prevents ftrace from instrumenting this function —
 * essential for production livepatch, harmless in skeleton mode.
 */
__attribute__((noinline, used))
u64 zenvecha_dummy_func(void)
{
	if (atomic_read(&zenvecha_dummy_patched))
		return ZENVECHA_DUMMY_PATCHED_VALUE;
	return ZENVECHA_DUMMY_ORIGINAL_VALUE;
}

/* Internal state setters — called only from the executor inside
 * stop_machine. */
void zenvecha_dummy_set_patched(bool patched)
{
	atomic_set(&zenvecha_dummy_patched, patched ? 1 : 0);
}

bool zenvecha_dummy_is_patched(void)
{
	return atomic_read(&zenvecha_dummy_patched) != 0;
}
